target/linux/ramips/dts/mt7621_keenetic_kn-3510.dts

```diff
@@ -0,0 +1,247 @@
+			color = <LED_COLOR_ID_AMBER>;
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		led_status_green: led-1 {
+			function = LED_FUNCTION_STATUS;
+			color = <LED_COLOR_ID_GREEN>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		wps {
+			label = "wps";
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
+		fn1 {
+			label = "fn1";
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+		};
+	};
+
+	virtual_flash {
+		compatible = "mtd-concat";
+		devices = <&firmware1 &storage1 &firmware2 &storage2>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "kernel";
+				reg = <0x0 0x400000>;
+			};
+
+			partition@400000 {
+				label = "ubi";
+				reg = <0x400000 0x0>;
+			};
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "uart3", "jtag";
+		function = "gpio";
+	};
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x80000>;
+			read-only;
+		};
+
+		partition@80000 {
+			label = "u-config";
+			reg = <0x80000 0x80000>;
+			read-only;
+		};
+
+		partition@100000 {
+			label = "rf-eeprom";
+			reg = <0x100000 0x80000>;
+			read-only;
+
+			nvmem-layout {
+				compatible = "fixed-layout";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				eeprom_factory_0: eeprom@0 {
+					reg = <0x0 0xe00>;
+				};
+
+				macaddr_factory_4: macaddr@4 {
+					reg = <0x4 0x6>;
+				};
+
+				macaddr_factory_a: macaddr@a {
+					reg = <0xa 0x6>;
+				};
+
+				precal_factory_e10: precal@e10 {
+					reg = <0xe10 0x19c10>;
+				};
+			};
+		};
+
+		firmware1: partition@180000 {
+			label = "firmware_1";
+			reg = <0x180000 0x1a40000>;
+		};
+
+		partition@1bc0000 {
+			label = "config_1";
+			reg = <0x1bc0000 0x20000>;
+			read-only;
+		};
+
+		partition@1dc0000 {
+			label = "storage_legacy";
+			reg = <0x1dc0000 0x20000>;
+			read-only;
+		};
+
+		partition@1fc0000 {
+			label = "dump";
+			reg = <0x1fc0000 0x40000>;
+			read-only;
+		};
+
+		storage1: partition@2000000 {
+			label = "storage_a";
+			reg = <0x2000000 0x1fc0000>;
+		};
+
+		partition@3fc0000 {
+			label = "u-state";
+			reg = <0x3fc0000 0x80000>;
+			read-only;
+		};
+
+		partition@4040000 {
+			label = "u-config_res";
+			reg = <0x4040000 0x80000>;
+			read-only;
+		};
+
+		partition@40c0000 {
+			label = "rf-eeprom_res";
+			reg = <0x40c0000 0x80000>;
+			read-only;
+		};
+
+		firmware2: partition@4140000 {
+			label = "firmware_2";
+			reg = <0x4140000 0x1a40000>;
+		};
+
+		partition@5b80000 {
+			label = "config_2";
+			reg = <0x5d00000 0x20000>;
+			read-only;
+		};
+
+		storage2: partition@5d80000 {
+			label = "storage_b";
+			reg = <0x5d80000 0x2200000>;
```

Comment:
Why is this 2.25MiB larger than storage_a?
Was this the case in the stock firmware as well?
I'm just wondering

Reply:
In the case of this device, this partitions are not used at all since it is an access point. There is similiar device Keenetic KN-1010 with a same flash. There are also different sizes storage_a and storage_b, but it doesn't matter since they are combined in the one even in stock firmware to Storage partition.

Reply:
I didn't know they publish their source code fully. I just realized this.
https://github.com/keenetic/keenetic-sdk/tree/4.01.C.7.0-1/target/linux
but mt7621 is still without dts files. (unlike mt7981 or mt7622)

anyway: I was wondering whether they were using Mediatek nmbm or something similar to detect bad blocks but that doesn't appear to be the case. (unless you saw something in the serial output of the bootloader).
Do you happen to know which filesystem they are using for storage in stock rom? jffs2, ubi, etc.?
It's curious to see Keenetic apply A/B partitioning for every device but not use it for all of them.

Some implementations of A/B partitioning in particular Mediatek and like the Asus RT-AX53U do have A/B partitioning but only in the way that if the bootloader fails to boot the OS (once, three times I don't remember) then it writes the OS from Partition B to A. (it cannot boot from B)
But checksums might be involved so it checks whether partition B is valid beforehand.
I just wanted to throw this in here sorry if it's more confusing than helping.

In general it looks very good how you did everything and concatted the partitions :)
great job

Reply:
> I didn't know they publish their source code fully. I just realized this.

No, there is no code there. Essentially, this SDK allows to disassemble the binary firmware, put something to it and assemble firmware back.

> Do you happen to know which filesystem they are using for storage in stock rom? jffs2, ubi, etc.?
```
[I] Jan  1 00:00:05 kernel: Registering UBI data partitions parser
[W] Jan  1 00:00:05 kernel: Concatenating MTD devices:
[W] Jan  1 00:00:05 kernel: (0): "Storage_A"
[W] Jan  1 00:00:05 kernel: (1): "Storage_B"
[W] Jan  1 00:00:05 kernel: into device "NDM combined UBI partition"
[I] Jan  1 00:00:05 kernel: 1 ndmubipart partitions found on MTD device NDM combined UBI partition
[I] Jan  1 00:00:05 kernel: mtd: creating 1 MTD partitions on "NDM combined UBI partition":
[I] Jan  1 00:00:05 kernel: 0x000000000000-0x0000041c0000 : "Storage"
[I] Jan  1 00:00:05 kernel: Merging storage partitions OK   
```